scripts/west_commands/sbom/ninja_build_extractor.py

```diff
@@ -0,0 +1,383 @@
+    ARCHIVE = 'archive'
+    OBJ = 'obj'
+    STAMP = 'stamp'
+    MISSING = 'missing'
+    OTHER = 'other'
+
+
+class NinjaBuildExtractor:
+    '''
+    Class for extracting list of input files for a specific ninja build direcory.
+    '''
+
+    build_dir: Path
+    deps: 'dict[set[str]]'
+
+    archives: 'list[BuildArchive]'
+    process_target_queue: 'list[tuple[str, BuildObject, BuildArchive]]'
+    cache: 'dict[str, set[str]]'
+    file_type_cache: 'dict[str, FileType]'
+    lock: Lock
+    target_details: 'dict[str, tuple[Path, FileType]]'
+    root: BuildArchive
+    archives: 'list[BuildArchive]'
+
+
+    def __init__(self, build_dir: Path):
+        '''
+        Initialize build directory input object that will fill "data" object and get
+        information from "build_dir".
+        '''
+        self.process_target_queue = list()
+        self.cache = dict()
+        self.file_type_cache = dict()
+        self.build_dir = Path(build_dir)
+        self.lock = Lock()
+        self.target_details = dict()
+        self.root = BuildArchive()
+        self.root.path = None
+        dummy_object = BuildObject()
+        dummy_object.parent = self.root
+        self.root.objects[''] = dummy_object
+        self.archives = []
+        deps_file_name = command_execute(args.ninja, '-t', 'deps', cwd=self.build_dir,
+                                         return_path=True)
+        self.parse_deps_file(deps_file_name)
+
+
+    def detect_file_type(self, file: Path) -> FileType:
+        '''
+        Simple detector for type of a file based on its header.
+        '''
+        str_file_name = str(file)
+        if str_file_name in self.file_type_cache:
+            return self.file_type_cache[str_file_name]
+        if not file.exists():
+            self.file_type_cache[str_file_name] = FileType.MISSING
+            return FileType.MISSING
+        with open(file, 'r', encoding='8859') as fd:
+            header = fd.read(16)
+        if header.startswith('!<arch>\n'):
+            self.file_type_cache[str_file_name] = FileType.ARCHIVE
+            return FileType.ARCHIVE
+        elif header.startswith('\x7FELF'):
+            self.file_type_cache[str_file_name] = FileType.OBJ
+            return FileType.OBJ
+        elif file.suffix.lower() == '.stamp':
+            self.file_type_cache[str_file_name] = FileType.STAMP
+            return FileType.STAMP
+        else:
+            self.file_type_cache[str_file_name] = FileType.OTHER
+            return FileType.OTHER
+
+
+    def parse_deps_file(self, deps_file_name: str):
+        '''
+        Reads all dependencies stored in the .ninja_deps file and sores it in self.deps dictionary.
+        '''
+        self.deps = dict()
+        target_line_re = re.compile(r'([^\s]+)\s*:\s*(#.*)?')
+        dep_line_re = re.compile(r'\s+(.*?)\s*(#.*)?')
+        empty_line_re = re.compile(r'\s*(#.*)?')
+        with open(deps_file_name, 'r') as fd:
```

Comment:
For `detect_file_type` function, encoding is required, because the file that we are opening there may be binary, so we have to force encoding that will not raise any exceptions for binary files. This case is different, we are opening text only file using platform's default encoding and it is ok to raise an exception if there is something wrong with the file.